### PR TITLE
Prometheus: a handful of new per-exchange and per-queue metrics

### PR DESCRIPTION
**Pull request by LoisSotoLopez**
## Proposed Changes

Solves the issue explained at https://github.com/rabbitmq/rabbitmq-server/issues/11217

- Defines 3 more `rabbit_core_metrics` tables, where per queue, per exchange and per queue/exchange metrics will be stored without including a channel ID on the key.
- Initializes those tables and configures `rabbit_core_metrics` to populate them on `channel_stats` calls, apart from populating those metrics per channel.
- Configures the `rabbit_mgmt` and `prometheus_rabbitmq_core` `_metrics_collector`s to collect those metrics.
- Provides a few tests in the `rabbit_prometheus_http_SUITE`

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist
- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All tests pass locally with my changes
- [ ] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [ ] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

### CODE REVIEW

**Comment by michaelklishin** [comment]
```
cd deps/rabbitmq_prometheus
gmake tests
```

on this branch fails for me with

```
rabbit_prometheus_http_SUITE > per_object_metrics > globally_configure_per_object_metrics_test
    #1. {error,
            {{assertEqual,
                 [{module,rabbit_prometheus_http_SUITE},
                  {line,413},
                  {expression,
                      "re : run ( Body , \"^rabbitmq_queue_get_ack_total{\" , [ { capture , none } , multiline ] )"},
                  {expected,match},
                  {value,nomatch}]},
             [{rabbit_prometheus_http_SUITE,per_object_metrics_test,2,
                  [{file,"rabbit_prometheus_http_SUITE.erl"},{line,413}]},
              {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
              {test_server,run_test_case_eval1,6,
                  [{file,"test_server.erl"},{line,1302}]},
              {test_server,run_test_case_eval,9,
                  [{file,"test_server.erl"},{line,1234}]}]}}

rabbit_prometheus_http_SUITE > per_object_endpoint_metrics > endpoint_per_object_metrics
    #1. {error,
            {{assertEqual,
                 [{module,rabbit_prometheus_http_SUITE},
                  {line,413},
                  {expression,
                      "re : run ( Body , \"^rabbitmq_queue_get_ack_total{\" , [ { capture , none } , multiline ] )"},
                  {expected,match},
                  {value,nomatch}]},
             [{rabbit_prometheus_http_SUITE,per_object_metrics_test,2,
                  [{file,"rabbit_prometheus_http_SUITE.erl"},{line,413}]},
              {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
              {test_server,run_test_case_eval1,6,
                  [{file,"test_server.erl"},{line,1302}]},
              {test_server,run_test_case_eval,9,
                  [{file,"test_server.erl"},{line,1234}]}]}}
```

and one more error that can be inspected by opening `logs/index.html` after the CT test run.

**Comment by LoisSotoLopez** [comment]
> ```
> cd deps/rabbitmq_prometheus
> gmake tests
> ```
> 
> on this branch fails for me with
> 
> ```
> rabbit_prometheus_http_SUITE > per_object_metrics > globally_configure_per_object_metrics_test
>     #1. {error,
>             {{assertEqual,
>                  [{module,rabbit_prometheus_http_SUITE},
>                   {line,413},
>                   {expression,
>                       "re : run ( Body , \"^rabbitmq_queue_get_ack_total{\" , [ { capture , none } , multiline ] )"},
>                   {expected,match},
>                   {value,nomatch}]},
>              [{rabbit_prometheus_http_SUITE,per_object_metrics_test,2,
>                   [{file,"rabbit_prometheus_http_SUITE.erl"},{line,413}]},
>               {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
>               {test_server,run_test_case_eval1,6,
>                   [{file,"test_server.erl"},{line,1302}]},
>               {test_server,run_test_case_eval,9,
>                   [{file,"test_server.erl"},{line,1234}]}]}}
> 
> rabbit_prometheus_http_SUITE > per_object_endpoint_metrics > endpoint_per_object_metrics
>     #1. {error,
>             {{assertEqual,
>                  [{module,rabbit_prometheus_http_SUITE},
>                   {line,413},
>                   {expression,
>                       "re : run ( Body , \"^rabbitmq_queue_get_ack_total{\" , [ { capture , none } , multiline ] )"},
>                   {expected,match},
>                   {value,nomatch}]},
>              [{rabbit_prometheus_http_SUITE,per_object_metrics_test,2,
>                   [{file,"rabbit_prometheus_http_SUITE.erl"},{line,413}]},
>               {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
>               {test_server,run_test_case_eval1,6,
>                   [{file,"test_server.erl"},{line,1302}]},
>               {test_server,run_test_case_eval,9,
>                   [{file,"test_server.erl"},{line,1234}]}]}}
> ```
> 
> and one more error that can be inspected by opening `logs/index.html` after the CT test run.

Yep. Should be fixed up right now. Wasn't expecting the core team to look at draft PRs. Won't open them any more if that's the case. :+1:

File: deps/rabbitmq_prometheus/src/collectors/prometheus_rabbitmq_core_metrics_collector.erl [hunk 1]
```diff
@@ -544,15 +563,20 @@ get_data(queue_metrics = Table, false, VHostsFilter) ->
...
+                           Table == queue_exchange_metrics;
                            Table == channel_queue_exchange_metrics;
                            Table == ra_metrics;
                            Table == channel_process_metrics ->
     Result = ets:foldl(fun
                   %% For queue_coarse_metrics
                   ({#resource{kind = queue, virtual_host = VHost}, _, _, _, _}, Acc) when is_map(VHostsFilter), map_get(VHost, VHostsFilter) == false ->
                                Acc;
+                  ({#resource{kind = queue, virtual_host = VHost}, _, _, _, _, _, _, _, _}, Acc) when is_map(VHostsFilter), map_get(VHost, VHostsFilter) == false ->
+                               Acc;
```
Comment (gomoripeti) [inline deps/rabbitmq_prometheus/src/collectors/prometheus_rabbitmq_core_metrics_collector.erl:579]:
```suggestion
                               Acc;
                  ({{#resource{kind = queue, virtual_host = VHost}, #resource{kind = exchange}}, _, _}, Acc) when is_map(VHostsFilter), map_get(VHost, VHostsFilter) == false ->
                               Acc;
```

```

File: deps/rabbitmq_prometheus/src/collectors/prometheus_rabbitmq_core_metrics_collector.erl [hunk 2]
```diff
@@ -579,6 +603,42 @@ get_data(Table, false, VHostsFilter) when Table == channel_exchange_metrics;
...
+    ets:foldl(fun
+        ({#resource{kind = exchange, virtual_host = VHost}, _, _, _, _, _} = Row, Acc) when
+            map_get(VHost, VHostsFilter)
+        ->
+            [Row | Acc];
+        (_Row, Acc) ->
+            Acc
+    end, [], Table);
+get_data(exchange_metrics, true, _VhostsFilter) ->
+    [];
```
Comment (gomoripeti) [inline deps/rabbitmq_prometheus/src/collectors/prometheus_rabbitmq_core_metrics_collector.erl:616]: this is the opposite of the intended behaviour. If `VHostsFilter` is undefined (`false`) then metrics from all vhosts should be included
After second thought this clauses is not needed as the last clause of `get_data` which just calls ` ets:tab2list(Table).` will cover what is needed

File: deps/rabbit/src/rabbit_core_metrics_gc.erl [hunk 3]
```diff
@@ -92,14 +92,17 @@ gc_leader_data(Id, Table, GbSet) ->
 gc_global_queues() ->
     GbSet = gb_sets:from_list(rabbit_amqqueue:list_names()),
     gc_process_and_entity(channel_queue_metrics, GbSet),
+    gc_process_and_entity(queue_counter_metrics, GbSet),
```
Comment (gomoripeti) [inline deps/rabbit/src/rabbit_core_metrics_gc.erl:95]:
```suggestion
    gc_entity(queue_counter_metrics, GbSet),
```

**Comment by michaelklishin** [comment]
@LoisSotoLopez @gomoripeti thank you. This will ship in 4.0 this fall.

I'd appreciate if your `rabbitmq-server` fork was updated a little bit more frequently than once every ≈ 4 weeks. Rebasing a triple digit number of commits can take a while.

**Comment by michaelklishin** [comment]
I'm afraid `deps/rabbitmq_prometheus` test suite reliably fails with this PR, even with the new tests excluded from mixed version runs that they cannot possibly pass (against a 3.13.x node) https://github.com/rabbitmq/rabbitmq-server/pull/11687.

**Pull request merged**